### PR TITLE
Networking V2: Allow floatingip associate with the fixed_ip

### LINKED ISSUE #646: Not possible to specify fixed_ip for openstack_networking_floatingip_associate_v2
Hi there,

in certain configurations you'll need to use a workaround when trying to associate a pre-allocated Floating IP to a Networking port. Now OpenStack CLI (help output below shortened!) requires the statement of a fixed <ip-address> when a single Port has more than one of them. This currently is not supported by Terraform.

```bash
λ openstack floating ip set -h
usage: openstack floating ip set [-h] --port <port>
                                 [--fixed-ip-address <ip-address>]
                                 <floating-ip>

Set floating IP Properties

  <floating-ip>         Floating IP to associate (IP address or ID)
  --port <port>         Associate the floating IP with port (name or ID)
  --fixed-ip-address <ip-address>
                        Fixed IP of the port (required only if port has
                        multiple IPs)
```

### Terraform Version
Terraform v0.11.11

### Affected Resource(s)
Please list the resources as a list, for example:
-  openstack_networking_floatingip_associate_v2

### Terraform Configuration Files
```hcl
# Bind Port to existing FIP but only use it if NO fixed_ip is being stated
resource "openstack_networking_floatingip_associate_v2" "association" {
  count = "${var.floating_ip != "" && var.fixed_ip == "" ? 1 : 0}"

  floating_ip = "${var.floating_ip}"
  #fixed_ip   = "${var.fixed_ip}" not supported, yet!
  port_id     = "${var.port_id}"
}
```

### Expected Behavior
Terraform should be enabled to support this feature of the OpenStack API.

### Actual Behavior
Can't create the resource and instead need to call as a workaround.

```bash
λ openstack floating ip set --port ${var.port_id} --fixed-ip-address ${var.fixed_ip} ${var.floating_ip}
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

### Important Factoids
This applies to OpenStack (Ocata) clouds using OpenContrail (v3.2!) with AVI Vantage (v17.2.12).

### PR DESCRIPTION
**Pull request by kayrus**
Resolves #646

### CODE REVIEW

**Comment by jtopjian** [comment]
@kayrus It looks like this PR is making the following changes:

1. Adding validation to the `openstack_networking_floatingip_v2` data source
2. Adding validation to the `openstack_compute_floatingip_v2` resource
3. Adding a new argument to `openstack_networking_floatingip_associate_v2` resource
4. Adding a few points of validation to `openstack_networking_floatingip_v2` resource

All changes are valid, but can these be split into separate PRs?

Considering the issues we've had in the past with new validation, we should make these individual PRs so they each have a changelog entry that points to a single commit.

**Comment by kayrus** [comment]
@jtopjian totally agree

**Comment by kayrus** [comment]
@jtopjian ready for review

File: openstack/resource_openstack_networking_floatingip_v2_test.go [hunk 1]
```diff
@@ -186,24 +196,70 @@ resource "openstack_networking_port_v2" "port_1" {
   admin_state_up = "true"
   network_id = "${openstack_networking_subnet_v2.subnet_1.network_id}"
 
-  fixed_ip {
+  fixed_ip = {
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_v2_test.go]: This needs reverted - it's not valid syntax for v0.12. It's a little confusing and there will be some upcoming documentation to help with this. 

Declarations which have a nested schema do not use `=`.

Declarations which do not have a schema (such as `TypeMap` of a set of key/values) will need `=`.

side note: I've already confirmed that all other syntax in this provider is correct for v0.12.

File: openstack/resource_openstack_networking_floatingip_associate_v2.go [hunk 2]
```diff
@@ -50,27 +56,27 @@ func resourceNetworkingFloatingIPAssociateV2Create(d *schema.ResourceData, meta
 
 	floatingIP := d.Get("floating_ip").(string)
 	portID := d.Get("port_id").(string)
+	fixedIP := d.Get("fixed_ip").(string)
 
 	fipID, err := networkingFloatingIPV2ID(networkingClient, floatingIP)
 	if err != nil {
-		return fmt.Errorf("Unable to get ID of openstack_networking_floatingip_v2: %s", err)
+		return fmt.Errorf("Unable to get ID of floatingip: %s", err)
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_associate_v2.go]: nit: The original error message was indeed incorrect, but part of the code cleanup is to ensure error messages are as precise as possible. How about something like:

```go
fmt.Errorf("Unable to get ID of openstack_networking_floatingip_associate_v2 floating_ip %s: %s", floatingIP, err)
```

File: openstack/resource_openstack_networking_floatingip_associate_v2.go [hunk 3]
```diff
@@ -50,27 +56,27 @@ func resourceNetworkingFloatingIPAssociateV2Create(d *schema.ResourceData, meta
...
+		PortID:  &portID,
+		FixedIP: fixedIP,
 	}
 
 	log.Printf("[DEBUG] openstack_networking_floatingip_associate_v2 create options: %#v", updateOpts)
 	_, err = floatingips.Update(networkingClient, fipID, updateOpts).Extract()
 	if err != nil {
-		return fmt.Errorf("Error associating openstack_networking_floatingip_v2 %s to openstack_networking_port_v2 %s: %s",
-			fipID, portID, err)
+		return fmt.Errorf("Error associating floatingip %s with port %s: %s", fipID, portID, err)
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_associate_v2.go]: nit: similar change here for the precise error message. How about:

```go
fmt.Errorf("Error associating openstack_networking_floatingip_associate_v2 floating_ip %s with port %s: %s", fipID, portID, err)
```

File: openstack/resource_openstack_networking_floatingip_associate_v2.go [hunk 4]
```diff
@@ -50,27 +56,27 @@ func resourceNetworkingFloatingIPAssociateV2Create(d *schema.ResourceData, meta
...
 	}
 
 	d.SetId(fipID)
 
-	log.Printf("[DEBUG] Created association between openstack_networking_floatingip_v2 %s and openstack_networking_port_v2 %s",
-		fipID, portID)
+	log.Printf("[DEBUG] Created association between floatingip %s and port %s", fipID, portID)
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_associate_v2.go]: nit: similar change here for the precise error message:

```go
log.Printf("[DEBUG] Created association between openstack_networking_floatingip_associate_v2 floating_ip %s and port %s", fipID, portID)
```

Alternatively, this line can be removed - success can be inferred by other log messages.

File: openstack/resource_openstack_networking_floatingip_associate_v2.go [hunk 5]
```diff
@@ -107,11 +140,10 @@ func resourceNetworkingFloatingIPAssociateV2Delete(d *schema.ResourceData, meta
 		PortID: new(string),
 	}
 
-	log.Printf("[DEBUG] openstack_networking_floatingip_v2 disassociating options: %#v", updateOpts)
+	log.Printf("[DEBUG] openstack_networking_floatingip_associate_v2 disassociating options: %#v", updateOpts)
 	_, err = floatingips.Update(networkingClient, d.Id(), updateOpts).Extract()
 	if err != nil {
-		return fmt.Errorf("Error disassociating openstack_networking_floatingip_v2 %s from openstack_networking_port_v2 %s: %s",
-			d.Id(), portID, err)
+		return fmt.Errorf("Error disassociating floatingip %s from port %s: %s", d.Id(), portID, err)
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_associate_v2.go]: nit: How about:

```go
fmt.Errorf("Error disassociating openstack_networking_floatingip_associate_v2 floating_ip %s with port %s: %s", d.Id(), portID, err)
```

File: openstack/resource_openstack_networking_floatingip_associate_v2_test.go [hunk 6]
```diff
@@ -137,3 +175,101 @@ resource "openstack_networking_floatingip_associate_v2" "fip_1" {
...
+resource "openstack_networking_router_v2" "router_1" {
+  name = "router_1"
+  external_gateway = "%s"
+}
+
+resource "openstack_networking_port_v2" "port_1" {
+  admin_state_up = "true"
+  network_id = "${openstack_networking_subnet_v2.subnet_1.network_id}"
+
+  fixed_ip = {
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_floatingip_associate_v2_test.go]: `s/=//g`

File: openstack/resource_openstack_networking_floatingip_v2_test.go (hunk 1 above)
Comment (kayrus) [reply openstack/resource_openstack_networking_floatingip_v2_test.go]: Indeed, it is confusing.

**Comment by kayrus** [comment]
@jtopjian kindly ping

**Pull request merged**